File: Leetcode-Problems/Majority Element.cpp
// class Solution {
// public:
//     int majorityElement(vector<int>& nums) {
//         unordered_map<int,int>mp;
//         for(int i:nums){
//             mp[i]++;
//         }
//         for(auto i:mp){
//             if(i.second>nums.size()/2){
//                 return i.first;
//             }
//         }
//         return 0;
//     }
// };

class Solution
{
public:
    int majorityElement(vector<int> &nums)
    {
        int res = 0;
        int majority = 0;

        for (int n : nums)
        {
            if (majority == 0)
            {
                res = n;
            }

            majority += n == res ? 1 : -1;
        }

        return res;
    }
};

File: Leetcode-Problems/Majority Element.java
class Solution {
    public int majorityElement(int[] nums) {
        HashMap<Integer, Integer> mp = new HashMap<>();
        for (int i : nums) {
            mp.put(i, mp.getOrDefault(i, 0) + 1);
        }
        int n = nums.length;
        for (Map.Entry<Integer, Integer> entry : mp.entrySet()) {
            if (entry.getValue() > (n / 2)) {
                return entry.getKey();
            }
        }
        return -1;
    }
}

class Solution {
    public int majorityElement(int[] nums) {
        Arrays.sort(nums);
        return nums[nums.length / 2];
    }
}

class Solution {
    public int majorityElement(int[] nums) {
        int candidate = 0, count = 0;
        for (int i : nums) {
            if (count == 0) {
                candidate = i;
            }
            if (i == candidate) {
                count++;
            } else
                count--;
        }
        return candidate;
    }
}